lib/image-stream/parse-partitions.js

```diff
@@ -0,0 +1,205 @@
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const Bluebird = require('bluebird');
+const MBR = require('mbr');
+const GPT = require('gpt');
+
+/**
+ * @summary Maximum number of bytes to read from the stream
+ * @type {Number}
+ * @constant
+ */
+const MAX_STREAM_BYTES = 8192;
+
+/**
+ * @summary Initial number of bytes read
+ * @type {Number}
+ * @constant
+ */
+const INITIAL_LENGTH = 0;
+
+/**
+ * @summary Initial block size
+ * @type {Number}
+ * @constant
+ */
+const INITIAL_BLOCK_SIZE = 512;
+
+/**
+ * @summary Maximum block size to check for
+ * @type {Number}
+ * @constant
+ */
+const MAX_BLOCK_SIZE = 4096;
+
+/**
+ * @summary Attempt to parse the GPT from various block sizes
+ * @function
+ * @private
+ * @param {Buffer} buffer - Buffer
+ * @returns {GPT|null}
+ * @example
+ * const gpt = detectGPT(buffer)
+ *
+ * if (gpt != null) {
+ *   // Has a GPT
+ *   console.log('Partitions:', gpt.partitions)
+ * }
+ */
+const detectGPT = (buffer) => {
+
+  let blockSize = INITIAL_BLOCK_SIZE;
+  let gpt = null;
+
+  // Attempt to parse the GPT from several offsets,
+  // as the block size of the image may vary (512,1024,2048,4096);
+  // For example, ISOs will usually have a block size of 4096,
+  // but raw images a block size of 512 bytes
+  while (blockSize <= MAX_BLOCK_SIZE) {
+    gpt = _.attempt(GPT.parse, buffer.slice(blockSize));
+    if (!_.isError(gpt)) {
+      return gpt;
+    }
+    blockSize += blockSize;
+  }
+
+  return null;
+
+};
+
+/**
+ * @summary Attempt to parse the MBR & GPT from a given buffer
+ * @private
+ * @param {Object} image - Image metadata
+ * @param {Buffer} buffer - Buffer
+ * @example
+ * parsePartitionTables(image, buffer)
+ *
+ * if (image.hasMBR || image.hasGPT) {
+ *   console.log('Partitions:', image.partitions)
+ * }
+ */
+const parsePartitionTables = (image, buffer) => {
+
+  const mbr = _.attempt(MBR.parse, buffer);
+  let gpt = null;
+
+  if (!_.isError(mbr)) {
+    image.hasMBR = true;
+    gpt = detectGPT(buffer);
+    image.hasGPT = !_.isNil(gpt);
+  }
+
+  // As MBR and GPT partition entries have a different structure,
+  // we normalize them here to make them easier to deal with and
+  // avoid clutter in what's sent to analytics
+  if (image.hasGPT) {
+    image.partitions = _.map(gpt.partitions, (partition) => {
+      return {
+        type: partition.type.toString(),
+        id: partition.guid.toString(),
+        name: partition.name,
+        firstLBA: partition.firstLBA,
+        lastLBA: partition.lastLBA,
+        extended: false
+      };
+    });
+  } else if (image.hasMBR) {
+    image.partitions = _.map(mbr.partitions, (partition) => {
+      return {
+        type: partition.type,
+        id: null,
+        name: null,
+        firstLBA: partition.firstLBA,
+        lastLBA: partition.lastLBA,
+        extended: partition.extended
+      };
+    });
+  }
+
+};
+
+/**
+ * @summary Attempt to read the MBR and GPT from an imagestream
+ * @description
```

Comment:
Can we move `@description` below `@public` and `@function` for consistency purposes?